### ISSUE TITLE
[BUG] FSM not compatible with disabled domain reload

### ISSUE THREAD

**Issue by lgarczyn** [unity-atoms/unity-atoms#402]
**Describe the bug**
Variables use OnEnable to reset themselves

However, Unity does not call OnEnable on scriptable objects when entering playmode IFF:

* the ScriptableObject is already loaded
* domain reload on play is disabled

**To Reproduce**
1. Disable Domain Reload
2. Create any variable
3. Change the variable value during play mode
4. Value will be preserved on next play mode

**Expected behavior**
Variables should reset to initial value for every play mode sessions

**Additional context**
My current hack
```cs
using UnityAtoms;
using UnityEditor;
using UnityEngine;

namespace Atoms.Editor.EnsureVarReset
{
  // ensure class initializer is called whenever scripts recompile
  [InitializeOnLoad]
  public static class EnsureVarReset
  {
    // register an event handler when the class is initialized
    static EnsureVarReset()
    {
      // Whenever the game enters or leaves playmode
      EditorApplication.playModeStateChanged += LogPlayModeState;
    }

    static void LogPlayModeState(PlayModeStateChange state)
    {
      // Reset each loaded variable
      foreach (AtomBaseVariable variable in ScriptableObject.FindObjectsOfType<AtomBaseVariable>())
      {
        // Reset the var
        variable.Reset();
      }
    }
  }
}
```

**Comment by soraphis** [comment]
which version of atoms are you using?

in 4.4. this #182 should be in place, which works by registering to the playModeStateChanges already

**Comment by soraphis** [comment]
I followed your reproduction steps. For me it works fine. 
Will close the issue. Feel free to re-open it when you have more information about it.

**Comment by lgarczyn** [comment]
I'm on 4.4.5

I only noticed the bug when trying to use FSMs, can you try to reproduce with an FSM instead ?

**Comment by soraphis** [comment]
okay, FSM are infact the problem.

there are multiple layers to this, as it seems:

- OnEnable in AtomVariable is not virtual, so the OnEnable method in the FSM actually overrides the OnEnable of the base class and has no way to invoke it.
- Even after adjusting this, there is a weird behaviour in the FSM that the transition seems to only work every second time (yeah, don't ask me yet why this is the case)

**Comment by lgarczyn** [comment]
> okay, FSM are infact the problem.
> 
> there are multiple layers to this, as it seems:
> 
> - OnEnable in AtomVariable is not virtual, so the OnEnable method in the FSM actually overrides the OnEnable of the base class and has no way to invoke it.
> - Even after adjusting this, there is a weird behaviour in the FSM that the transition seems to only work every second time (yeah, don't ask me yet why this is the case)
>     - edit: I enabled domain reload and disabled it again, and now this bug disappeared 🙈 
> 

Glad to see the behavior is so weird for you too.

The bug doesn't happen every time, but the weirdest part is that the fsm is not even reset by the domain reload caused by compilation! (This might be because it's in its own assembly that's not recompiling)

**Comment by soraphis** [comment]
> but the weirdest part is that the fsm is not even reset by the domain reload caused by compilation! 

because it does not call the OnEnable of the base class. Already fixed this on my working branch

**Comment by 0xF6** [comment]
Hm, i encountered with this problem, in outside editor FSM has been incorrect state by default, not initial value :\
Fixed by enabling the inspector's debug mode and changing the value to initial

I have spent a lot of time searching for the reason why my builds are not working properly outside the editor
This is to the topic of #323 tasks, it would not be bad to have any notification in this case

**Comment by soraphis** [comment]
what is considered "incorrect state"? that's business logic, only you can decide about that and you could write an validator for that.

modifying the FSM implementation to throw errors in that case, would force you to setup every possible transition.

let's say you feed your keyboard input as actions into your FSM to handle your game state
WORLD ---ESC---> MENU
WORLD --- I ---> INVENTORY
...

that way you only handle transitions and every key that is not a transition is a "stay in state". works perfectly fine, but does not work anymore (or would spam warnings) if #323 would be implemented.

the issue in this case is less that the FSM does not throw warnings, the issue is more that the FSM was able to be in an invalid starting state that was seemingly not visible in the inspector (as you had to resolve to the debug inspector)

**Comment by 0xF6** [comment]
>but does not work anymore (or would spam warnings)

Not necessarily a warning, just enough any diagnostic logs

**Comment by soraphis** [comment]
> Not necessarily a warning, just enough any diagnostic logs

if you wanna elaborate on that topic or bring up implementation ideas that could solve that without harming performance or other implementations, feel free to do so, But I'd ask you, to do so in #323 (feel free to re-open the ticket if necessary!)

**Comment by AdamRamberg** [comment]
Closing issue. Fixed in #422
